lib/instrumentation/http-shared.js

```diff
@@ -49,9 +50,12 @@ exports.instrumentRequest = function (agent, moduleName) {
 
             // Handle case where res.end is called after an error occurred on the
             // stream (e.g. if the underlying socket was prematurely closed)
-            res.on('prefinish', function () {
+            const end = res.end
+            res.end = function () {
+              const result = end.apply(this, arguments)
               trans.end()
```

Comment:
@watson doesn't this assume that `prefinish` will be emitted synchronously? Did I misunderstand your comment https://github.com/elastic/apm-agent-nodejs/issues/1411#issuecomment-538927796?

> I guess we could instead listen for when someone calls response.end() and then end the transaction then. That would get around the stream related issues, but that might then wrongly end a transaction before the actual HTTP response is ended as any buffering happening internally in the response object after calling response.end() would be ignored (which is why we made our current implementation the way we did).

Won't this change end the transaction after buffering the data, but before the data is actually sent? That's why I thought we would keep listening for `prefinish`, and only end the transaction in `res.end` if `req.aborted === true`.

Reply:
No, this PR only uses the call to `response.end()` as a signal to end the transaction in case the underlying socket was closed prematurely - so as you said, any buffered data would be lost anyway. Under normal circumstances, we wait until the stream closes and then end the transaction. This is determined based on whether or not [the callback](https://github.com/elastic/apm-agent-nodejs/blob/37524b1a25d17c8cccc85a77c955cd29350206a4/lib/instrumentation/http-shared.js#L37) to `endOfStream` is called with an error or not.

Previously we only listened for the `prefinish` event in case we detected that the underlying socket closed prematurely. But the problem was that the `prefinish` event wasn't always emitted (if the race condition was triggered).

This PR instead relies on [`response.end()`](https://github.com/nodejs/node/blob/064e111515185529b6f943dc66929440557fd609/lib/_http_outgoing.js#L698-L766) being called in the cases where the stream encountered an error (like the socket being closed prematurely). Normally the `prefinish` event is emitted [right before](https://github.com/nodejs/node/blob/064e111515185529b6f943dc66929440557fd609/lib/_http_outgoing.js#L762) the `end` function runs to completion, so this shouldn't really change that behavior, except that we now always end the transaction, instead of only if the `prefinish` event was emitted.

Reply:
> This is determined based on whether the callback to endOfStream is called with an error or not.

Got it - I missed this fact. So if and only if the stream ends due to an error, then we end the transaction on `res.end`. Thanks for the explanation!